### PR TITLE
add nested portable version test

### LINKED ISSUE #288: Portable Serialization incompatible class definition when nested portables used
See original issue on java https://github.com/hazelcast/hazelcast/issues/12733

### PR DESCRIPTION
**Pull request by furkansenharputlu**
This pr implements `ClassDefinitionBuilder` and fixes #288.

### CODE REVIEW

File: src/serialization/portable/ClassDefinition.ts [hunk 1]
```diff
@@ -18,7 +18,7 @@ export class ClassDefinition {
     private factoryId: number;
     private classId: number;
     private version: number;
-    private fields: {[name: string]: FieldDefinition} = {};
+    private fields: { [name: string]: FieldDefinition } = {};
```
Comment (mustafaiman) [inline src/serialization/portable/ClassDefinition.ts]: add checking fields to `equals` method of `ClassDefinition`

File: src/serialization/portable/ClassDefinitionContext.ts [hunk 2]
```diff
@@ -55,9 +52,11 @@ export class ClassDefinitionContext {
             this.classDefs[cdKey] = classDefinition;
             return classDefinition;
         }
-        if (current instanceof ClassDefinition && !current.equals(classDefinition)) {
-            throw new RangeError(`Incompatible class definition with same class id: ${classDefinition.getClassId()}`);
+        if (!deepEqual(current, classDefinition)) {
```
Comment (mustafaiman) [inline src/serialization/portable/ClassDefinitionContext.ts]: use `ClassDefinition`s `equals` method for this check.

File: package.json [hunk 3]
```diff
@@ -5,6 +5,7 @@
   "main": "lib/index.js",
   "dependencies": {
     "bluebird": "3.5.0",
+    "deep-equal": "^1.0.1",
```
Comment (mustafaiman) [inline package.json]: We do not need a dependency for such a simple task.
If you need new generic functionality, first check if a method for that exists in `Util.ts`. If not, add the required method in there.

File: src/HazelcastError.ts [hunk 4]
```diff
@@ -32,6 +33,13 @@ export class HazelcastError extends Error {
     }
 }
 
+export class HazelcastSerializationError extends HazelcastError {
```
Comment (mustafaiman) [inline src/HazelcastError.ts]: also add this error to `ErrorFactory.ts`.

File: src/serialization/portable/ClassDefinitionBuilder.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+    addUTFField(fieldName: string): ClassDefinitionBuilder {
+        return this.addField(new FieldDefinition(this.index, fieldName, FieldType.UTF, this.version));
+    }
+
+    addPortableField(fieldName: string, def: ClassDefinition): ClassDefinitionBuilder {
+        if (def.getClassId() === 0) {
+            throw new RangeError('Portable class ID cannot be zero!');
+        }
+
+        return this.addField(new FieldDefinition(this.index, fieldName, FieldType.PORTABLE, this.version,
```
Comment (mustafaiman) [inline src/serialization/portable/ClassDefinitionBuilder.ts]: should `this.version` be `def.getVersion()`?

File: src/serialization/portable/ClassDefinitionBuilder.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    addUTFArrayField(fieldName: string): ClassDefinitionBuilder {
+        return this.addField(new FieldDefinition(this.index, fieldName, FieldType.UTF_ARRAY, this.version));
+    }
+
+    addPortableArrayField(fieldName: string, def: ClassDefinition): ClassDefinitionBuilder {
+        if (def.getClassId() === 0) {
+            throw new RangeError('Portable class ID cannot be zero!');
+        }
+
+        return this.addField(new FieldDefinition(this.index, fieldName, FieldType.PORTABLE_ARRAY, this.version,
```
Comment (mustafaiman) [inline src/serialization/portable/ClassDefinitionBuilder.ts]: `this.version` -> `def.getVersion()`

File: src/serialization/portable/ClassDefinitionBuilder.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    build(): ClassDefinition {
+        this.done = true;
+        var cd = new ClassDefinition(this.factoryId, this.classId, this.version);
+        for (let fd of this.fieldDefinitions) {
+            cd.addFieldDefinition(fd);
+        }
+        return cd;
+    }
+
+}
```
Comment (mustafaiman) [inline src/serialization/portable/ClassDefinitionBuilder.ts]: can you add `getFactoryId()`, `getClassId()`, `getVersion()` methods?

File: src/serialization/portable/PortableSerializer.ts [hunk 8]
```diff
@@ -73,7 +69,7 @@ export class PortableSerializer implements Serializer {
                 input.position(backupPos);
             }
         }
-        var reader: PortableReader;
+        var reader: DefaultPortableReader;
```
Comment (mustafaiman) [inline src/serialization/portable/PortableSerializer.ts]: this change is unnecessary. `PortableReader` is the interface. Any `reader` that is `PortableReader` is suitable here.

File: src/serialization/portable/ClassDefinitionBuilder.ts [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {ClassDefinition, FieldDefinition, FieldType} from './ClassDefinition';
+import {HazelcastSerializationError} from '../../HazelcastError';
+
+export class ClassDefinitionBuilder {
```
Comment (mustafaiman) [inline src/serialization/portable/ClassDefinitionBuilder.ts]: this class should be exported in `index.ts` for public use

File: src/serialization/portable/ClassDefinitionContext.ts [hunk 10]
```diff
@@ -48,9 +51,13 @@ export class ClassDefinitionContext {
             this.classDefs[cdKey] = classDefinition;
             return classDefinition;
         }
-        if (current instanceof ClassDefinition && !current.equals(classDefinition)) {
-            throw new RangeError(`Incompatible class definition with same class id: ${classDefinition.getClassId()}`);
+        try {
+            deepEqual(current, classDefinition);
```
Comment (mustafaiman) [inline src/serialization/portable/ClassDefinitionContext.ts]: I assume `deepEqual` checks private fields that are supposed to be implementation details. This is not good practice. A good way would be to have a public `equals` method in `ClassDefinition` and have it do the comparison.

File: src/serialization/portable/ClassDefinitionBuilder.ts (hunk 7 above)
Comment (mustafaiman) [reply src/serialization/portable/ClassDefinitionBuilder.ts]: I still do not see these methods.

File: src/serialization/portable/PortableSerializer.ts [hunk 11]
```diff
@@ -73,7 +69,7 @@ export class PortableSerializer implements Serializer {
                 input.position(backupPos);
             }
         }
-        let reader: PortableReader;
+        let reader: DefaultPortableReader;
```
Comment (mustafaiman) [inline src/serialization/portable/PortableSerializer.ts]: Why change to `DefaultPortableReader` from `PortableReadier?

Comment (furkansenharputlu) [reply src/serialization/portable/PortableSerializer.ts]: `PortableReader` doesn't have `end` function so I needed to use `DefaultPortableReader`.

Comment (mustafaiman) [reply src/serialization/portable/PortableSerializer.ts]: 👍

**Pull request merged**